### PR TITLE
Support string escapes parsing

### LINKED ISSUE #4: Support for unicode escapes
Unicode symbols represented as UTF-8 in strings are parsed as expected. I also see that parser has state `ES` (escape). But escapes of Unicode code-points are not converted to respective code points.
For example source text `"\"\\u00e9\"".getBytes()` is parsed as single token `"\\u00e9"` instead of `"é"`.

Here's the test ba58753276966227496d447dff6e9f0806b3a841

### PR DESCRIPTION
**Pull request by PetrGlad**
Implementation of code point parser and single-char escapes, fixes #4 . Add code to handle "escape" events which previously just changed state. 

Use Jackson for pretty-printing JSON generation. To correctly render escaped sequences existing pretty printer should have been fixed but I decided not to fix it and use Jackson instead.

Extract common example/test code into separate class. Also I extracted feeding/event reading loop into separate class (`de.undercouch.actson.JsonHelper`). 

Use named constants for escape actions. To keep formatting of transition table since now actions take 3 positions. I'd use named constants for other actions but that probably should be in separate PR.

Results of tests https://github.com/nst/JSONTestSuite do not seem to be changed compared to master.

### CODE REVIEW

**Comment by michel-kraemer** [comment]
Thanks! I had a quick look at your PR and it seems to be very good. I'll review it more thoroughly later and get back to you.

File: build.gradle [hunk 1]
```diff
@@ -24,6 +24,7 @@ tasks.withType(JavaCompile) {
 }
 
 dependencies {
+    testCompile 'com.fasterxml.jackson.core:jackson-core:2.8.4'
```
Comment (michel-kraemer) [inline build.gradle]: It should not be necessary to add this dependency explicitly. jackson-databind should already depend on it.

File: src/main/java/de/undercouch/actson/JsonParser.java [hunk 2]
```diff
@@ -399,14 +402,40 @@ private void parse(char nextChar) {
         event1 = stateToEvent();
       }
 
-      // Change the state.
```
Comment (michel-kraemer) [inline src/main/java/de/undercouch/actson/JsonParser.java]: Why did you delete this comment? Anything wrong with it?

File: src/main/java/de/undercouch/actson/JsonParser.java [hunk 3]
```diff
@@ -399,14 +402,40 @@ private void parse(char nextChar) {
...
       state = nextState;
+    } else if (nextState == XU) {
```
Comment (michel-kraemer) [inline src/main/java/de/undercouch/actson/JsonParser.java]: This whole block (and the following) should be moved to `performAction`.

File: src/main/java/de/undercouch/actson/JsonParser.java [hunk 4]
```diff
@@ -399,14 +402,40 @@ private void parse(char nextChar) {
...
+      // End of UTF code point escape
+      currentValue.append(nextChar);
+      final int len = currentValue.length();
```
Comment (michel-kraemer) [inline src/main/java/de/undercouch/actson/JsonParser.java]: `final` is unnecessary

File: src/main/java/de/undercouch/actson/JsonParser.java [hunk 5]
```diff
@@ -399,14 +402,40 @@ private void parse(char nextChar) {
...
+      final int code = Integer.parseInt(currentValue.substring(len - 4, len), 16);
```
Comment (michel-kraemer) [inline src/main/java/de/undercouch/actson/JsonParser.java]: `final` is unnecessary

File: src/test/java/de/undercouch/actson/PrettyPrinter.java [hunk 6]
```diff
@@ -23,12 +23,11 @@
 
 package de.undercouch.actson;
 
-import java.nio.charset.StandardCharsets;
```
Comment (michel-kraemer) [inline src/test/java/de/undercouch/actson/PrettyPrinter.java]: Please do not change this class too much. It's an example and the README links to it. It should be self-contained and have no dependencies (to Jackson).

File: src/test/java/de/undercouch/actson/JsonParserTest.java [hunk 7]
```diff
@@ -189,6 +180,28 @@ public void emptyObject() {
     assertJsonObjectEquals(json, parse(json));
   }
 
+  @Test
```
Comment (michel-kraemer) [inline src/test/java/de/undercouch/actson/JsonParserTest.java]: Please add comments to your new test methods. All other test methods have comments.

File: build.gradle (hunk 1 above)
Comment (PetrGlad) [inline build.gradle]: But that dependency is not explicit it just so happened that this library is in class path.
This now declares that we do directly require o.f.j.core functions in the code.

File: src/main/java/de/undercouch/actson/JsonParser.java (hunk 2 above)
Comment (PetrGlad) [inline src/main/java/de/undercouch/actson/JsonParser.java]: It states obvious. There would be some merit in it if this was unusual way or place to change the state. Neither is the case here. Also I do the same just below it twice more so it would cause additional confusion if some assignments had this comment while others - don't.

File: src/test/java/de/undercouch/actson/PrettyPrinter.java (hunk 6 above)
Comment (PetrGlad) [inline src/test/java/de/undercouch/actson/PrettyPrinter.java]: My motivation was that since I need correct serialization of escaped chars back to JSON I'd need to modify this printer accordingly which seemed like unnecessary work since it's just demo and test code - not the core functionality of Actson. Also I think that it is a good idea to not have duplicating functionality in demo and test code. 

That said, this version is required for tests to run correctly. If you want to still have have demo code without extra dependencies then probably I should keep older version as a separate class (maybe with caveat that it does not handle escapes).

File: src/test/java/de/undercouch/actson/JsonParserTest.java (hunk 7 above)
Comment (PetrGlad) [inline src/test/java/de/undercouch/actson/JsonParserTest.java]: I thought that method names are descriptive enough. I'll update this.

File: src/main/java/de/undercouch/actson/JsonParser.java (hunk 3 above)
Comment (PetrGlad) [inline src/main/java/de/undercouch/actson/JsonParser.java]: This would require to add nextChar to performAction parameters. 
Like `private void performAction(byte action, nextChar) { ... }`
Is this OK?

Comment (PetrGlad) [inline src/main/java/de/undercouch/actson/JsonParser.java]: Also currently no code in `performAction` modifies `this.currentValue`. So these sections that update `this.currentValue` seemed to me like a gray area :)

**Comment by PetrGlad** [comment]
OK, I'll update this to your requirements, there's nothing that cannot be fixed later.
